CRM/PivotReport/DataActivity.php

```diff
@@ -0,0 +1,374 @@
+<?php
+
+/**
+ * Provides a functionality to prepare entity data for Pivot Table.
+ */
+class CRM_PivotReport_DataActivity extends CRM_PivotReport_AbstractData {
+
+  public function __construct($name = NULL) {
+    $this->name = 'Activity';
+  }
+
+  /**
+   * Rebuilds pivot report cache including header and data.
+   *
+   * @param \CRM_PivotCache_AbstractGroup $cacheGroup
+   * @param array $params
+   *
+   * @return array
+   */
+  public function rebuildCache($cacheGroup, array $params) {
+    $this->fields = $this->getFields();
```

Comment:
If I'm not wrong, you're calling `getFields()` here because you need the to use `$this->fields` in `getEmptyRow` right? If that is the case, then the `getFields()` call should be inside of that method, and not here. If you did this for performance reasons,  you can update `getFields()` to cache the date it needs. Something like this:

```php
public function getFields() {
    if(empty($this->fields)) {
       // huge calculation to get fields
       $this->fields = $fields;
    }

    return $this->fields;
}
```

And then you can always use `$this->getFields()` whenever you want